### PR TITLE
Remove project usages in internal tasks for better configuration cache support

### PR DESCRIPTION
**Pull request by breskeby**
- This brings us closer to be compliant with the gradle configuration cache feature
   (see https://github.com/elastic/elasticsearch/issues/57918)
- All internal tasks used during precommit should now be free of invalid project references

### CODE REVIEW

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/LicenseHeadersTask.groovy [hunk 1]
```diff
@@ -178,4 +180,7 @@ class LicenseHeadersTask extends AntTask {
             throw new IllegalStateException("License header problems were found! Full details: " + reportFile.absolutePath)
         }
     }
+
+    @Internal
+    abstract ListProperty<FileCollection> getSourceFolders();
```
Comment (mark-vieira) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/LicenseHeadersTask.groovy:185]: Should we be using these "managed" properties more often?

File: buildSrc/src/main/java/org/elasticsearch/gradle/precommit/ForbiddenPatternsTask.java [hunk 2]
```diff
@@ -164,4 +173,11 @@ public void rule(Map<String, String> props) {
         // TODO: fail if pattern contains a newline, it won't work (currently)
         patterns.put(name, pattern);
     }
+
+    @Internal
+    abstract ListProperty<FileTree> getSourceFolders();
+
+    void setRootDir(File rootDir) {
```
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/precommit/ForbiddenPatternsTask.java:180]: Why not use a `DirectoryProperty` here?

File: buildSrc/src/main/java/org/elasticsearch/gradle/precommit/PrecommitTask.java [hunk 3]
```diff
@@ -31,12 +33,16 @@
...
+        return new File(getProjectLayout().getBuildDirectory().getAsFile().get(), "markers/" + this.getName());
     }
 
     @TaskAction
     public void writeMarker() throws IOException {
         Files.write(getSuccessMarker().toPath(), new byte[] {}, StandardOpenOption.CREATE);
     }
 
+    @Inject
+    protected ProjectLayout getProjectLayout() {
```
Comment (mark-vieira) [inline buildSrc/src/main/java/org/elasticsearch/gradle/precommit/PrecommitTask.java:45]: Where did we land on using method injection vs constructor injection? Should we standardize on one vs the other?

**Review by mark-vieira — APPROVED** [review]
Some minor comments but otherwise LGTM.

Also, any idea why `TestingConventionsTasks` is plural?

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/LicenseHeadersTask.groovy (hunk 1 above)
Comment (breskeby) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/LicenseHeadersTask.groovy:185]: I think it makes sense to move our build towards using these

File: buildSrc/src/main/java/org/elasticsearch/gradle/precommit/PrecommitTask.java (hunk 3 above)
Comment (breskeby) [reply buildSrc/src/main/java/org/elasticsearch/gradle/precommit/PrecommitTask.java:45]: I think by default we agreed on ctor injection. In this case I went with method injection due to keeping this local as this is an abstract task and all its descends would need to pass the injected service down here

**Comment by breskeby** [comment]
There seems to be quite an performance regression with these changes. I need to dig into this before we can merge that

**Comment by breskeby** [comment]
@elasticmachine test this please

**Comment by breskeby** [comment]
@elasticmachine test this please

**Comment by breskeby** [comment]
@elasticsearchmachine test this please

**Pull request merged**